Support target prefix with JSON format

hoangcuong2011 opened the pull request:
This feature allows Sockeye to add target prefix and target prefix factors during inference with JSON format. During inference target prefix can be specified with JSON format as follows:

```json
{ "text": "The boy ate the waff@@ le .", "target_prefix": "2XX"}
```

If a model was trained with target factors, we can add target prefix factors during inference with JSON format as follows:

```json
{ "text": "The boy ate the waff@@ le .", "target_prefix_factors": ["O"]}
```

Meanwhile, we can also add both target prefix and target prefix factors at the same time with JSON format, e.g.,:

```json
{ "text": "The boy ate the waff@@ le .", "target_prefix": "2XX", "target_prefix_factors": ["O"]}
```

Note that if an input is very long, Sockeye chunks the text and translates each chunk separately. By default, target prefix and target prefix factors are added to all chunks in that case. Alternatively, we can set `use_target_prefix_all_chunks` to `false` to add them only to the first chunk, e.g.,:

```json
{ "text": "The boy ate the waff@@ le .", "target_prefix": "2XX", "target_prefix_factors": ["O"], "use_target_prefix_all_chunks": false}
```

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mjdenkowski commented on sockeye/inference.py:
```diff
@@ -183,13 +197,18 @@ def chunks(self, chunk_size: int) -> Generator['TranslatorInput', None, None]:
             constraints = self.constraints if chunk_id == 0 else None
+            # Target_prefix_tokens are also assigned only to the first chunk
+            target_prefix_tokens = self.target_prefix_tokens if chunk_id == 0 else None
+            keep_target_prefix_key = self.keep_target_prefix_key if chunk_id == 0 else True
```
Good call on stating this explicitly!

This looks like something else we want to configure in the JSON input.  For example, if we're continuing a partial translation, we only want to apply the prefix to the first chunk.  If we're using special target tokens for multilingual NMT, we need to add them to all chunks so we know which language we're translating into.

hoangcuong2011 replied:
Done, thanks!

mjdenkowski reviewed (commented):
Thanks @hoangcuong2011!

A few comments:

- Having an option to automatically remove target prefix tokens is a great idea!  It avoids the need for users to write their own post-processing scripts.

- It looks like we also need an option to specify whether to add prefix tokens to just the first chunk or all chunks.

- It's great to see a unit test for `test_adjust_first_step_masking()`!  Can we add a test to the end of test_beam_search.py that includes specifying a target prefix?

fhieber commented on sockeye/beam_search.py:
```diff
@@ -864,6 +880,87 @@ def forward(self,
+
+        target prefix     initialize masking     masking           roll one step to the right
+                          from target prefix     is not 0           and assign 1 at index 0
+           [1 2]    ->       [1 2 0]     ->      [1 1 0]                   [1 1 1]
```
could you explain why we need the roll to the right?

hoangcuong2011 replied:
Hi Felix, I think it is best to explain it this way: 

If there is no prefix, we add first_step 1 time
If there is a prefix with length 1, we add first step 2 times
If there is a prefix with length 2, we add first step 3 times
If there is a prefix with length 3, we add first step 4 times
...

So the number of times we add first_step is equal to the length of prefix + 1. Rolling to the right reflects this.

fhieber commented on sockeye/inference.py:
```diff
@@ -303,6 +323,10 @@ def make_input_from_dict(sentence_id: SentenceId,
+        target_prefix_tokens = input_dict.get(C.JSON_TARGET_PREFIX_KEY)
+        target_prefix_tokens = list(utils.get_tokens(target_prefix_tokens)) if target_prefix_tokens else None
+        keep_target_prefix_key = input_dict.get(C.JSON_KEEP_TARGET_PREFIX_KEY)
+        keep_target_prefix_key = False if (keep_target_prefix_key is not None and keep_target_prefix_key == "False") else True
```
would the string value of `keep_target_prefix_key` be `"False"` or `"false"` or `False` ? I think we should follow JSON convention, see my comment above.

hoangcuong2011 replied:
Done. Thanks!

fhieber commented on docs/inference.md:
````diff
@@ -96,6 +96,18 @@ Similar to source factors, source prefix factors can be also specified with JSON
+By default, the translation output includes the target prefix as its first string. Alternatively, you can remove the target prefix from the translation output by setting `keep_target_prefix` to be `False`, e.g.,:
+
+```json
+{ "text": "The boy ate the waff@@ le .", "target_prefix": "2XX", "keep_target_prefix": "False"}
````
Why encode the boolean as a string? I think we should use proper JSON booleans: https://stackoverflow.com/questions/44616911/why-does-json-dumps-change-the-casing-of-boolean-value/44616947

hoangcuong2011 replied:
Done, thanks!

fhieber commented on sockeye/beam_search.py:
```diff
@@ -795,15 +807,19 @@ def forward(self,
                 # of the first row only by setting all other rows to inf
-                if t == 1:
-                    scores += first_step_mask
+                if t <= first_step_mask.size(-1):
```
This is critical path code for performance and this change will request the size of a tensor repeatedly, as well as require additional indexing on `first_step_mask` below.
Can we change the logic to keep the existing code for `target_prefix == None`?

It is a bit unfortunate that we are adding more data-dependent control flow here, it doesn't easily allow us to fold this into a traced module.

fhieber reviewed (commented):
Thanks for adding this! I have a few comments, some more high level:

1. The current implementation effectively constraints the first t beam search steps to the target prefix, which is neat and probably sufficient for most use cases where the target prefix is short. I wonder though if there would be value in thinking about how we could implement this with code from the scoring modules, the advantage being that computation over time steps within the target prefix could be parallelized. This would likely be a bit more involved (also because of JIT compilation etc.), but just wanted to hear your thoughts on this.
2. Since you are technically scoring the target prefix, can you create an integration/system test that asserts on output and score equality between: (a) scoring some data and (b) translating the source-sides data with the target-side encoded as target prefixes (including `</s>`?)? The existing integration/system tests already do this to some extent: we score the inputs and outputs from a previous beam search and assert that the scores are equal: https://github.com/awslabs/sockeye/blob/ea081433d0058c646da247412812c421012c0e43/test/common.py#L108

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -532,12 +532,14 @@ def __init__(self,
                  num_target_factors: int,
-                 inference: _SingleModelInference):
+                 inference: _SingleModelInference,
+                 output_vocab_size: Optional[int] = -1):
```
Done. Thanks!

hoangcuong2011 replied:
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -589,6 +593,10 @@ def forward(self,
             scores, model_states, target_factors = self._inference.decode_step(best_word_index,
                                                                                model_states,
                                                                                vocab_slice_ids=vocab_slice_ids)
+            if target_prefix is not None and t <= target_prefix.size(-1):
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -795,15 +807,19 @@ def forward(self,
             scores, lengths = self._update_scores(target_dists, finished, scores_accumulated,
                                                   lengths, max_output_lengths, pad_dist, eos_dist)
 
+            if target_prefix is not None and t <= target_prefix.size(-1):
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -864,6 +880,87 @@ def forward(self,
+                            [1 1 0]                        [inf, inf, 0]
+                            [1 1 0]                        [inf, inf, 0]
+                            [1 1 0]                        [inf, inf, 0]
+        """
```
Done. Thanks!
Great catch on Step 2 btw!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -589,6 +593,10 @@ def forward(self,
                                                                                model_states,
                                                                                vocab_slice_ids=vocab_slice_ids)
+            if target_prefix is not None and t <= target_prefix.size(-1):
+                target_prefix_in_one_hot = utils.one_hot_encoding_from_target_prefix(target_prefix[:,t-1:t], self.output_vocab_size).squeeze(1)
```
Done. You are right, it needs to change a bit with vocabulary selection. I fixed it. Thanks!

hoangcuong2011 commented:
Thank you for your great feedbacks. I revised the code to address all of them. Please review it again.

@mjdenkowski:
- It looks like we also need an option to specify whether to add prefix tokens to just the first chunk or all chunks. -> Done
- It's great to see a unit test for test_adjust_first_step_masking()! Can we add a test to the end of test_beam_search.py that includes specifying a target prefix -> Done. I added the test.

@fhieber 
- Since you are technically scoring the target prefix, can you create an integration/system test that asserts on output and score equality between: (a) scoring some data and (b) translating the source-sides data with the target-side encoded as target prefixes (including </s>?)? The existing integration/system tests already do this to some extent: we score the inputs and outputs from a previous beam search and assert that the scores are equal:
sockeye/test/common.py -> DONE. I added the test to make sure scoring is correct.

- Do we need to add support for target factors? -> DONE. I added new features that support target factors as well (plus different tests to make sure it corrects).


About Felix's point: ``I wonder though if there would be value in thinking about how we could implement this with code from the scoring modules, the advantage being that computation over time steps within the target prefix could be parallelized." This is a great thing to think of indeed. I thought about it. The good thing is as you mention we can parallel the computation. However, there are also important disadvantages:

- Implementing by constraining the beam like what we are doing matches with the way we are implementing decoding. Thus the code is easy to read and understand. If we implement in the parallelism way, it breaks the matching in the logic of the implementation to some degree I think.

- The length of prefix in MT is not as long as the prefix that is used, for instance, in GPT3. I think for MT it is about 0-1-2 tokens at most. So the benefit of parallelism could be marginal (but let's see how our ``customers" use the features, maybe they want to use long prefix)

- By implementing like we are doing, our code is very flexible to handle sentences in the same batch with different target prefixes (e.g. one has zero prefix, one has one prefix token, one has 2 prefix token). Implementing stuff in parallel makes it very nontrivial to handle that flexibility.

hoangcuong2011 commented on sockeye/test_utils.py:
```diff
@@ -157,8 +203,12 @@ def tmp_digits_dataset(prefix: str,
                 data['train_target_factors'].append(train_factor_path)
                 data['dev_target_factors'].append(dev_factor_path)
-                data['test_target_factors'].append(dev_factor_path)
+                data['test_target_factors'].append(test_factor_path)
```
btw I fixed this bug from previous code (dev_factor_path was used instead of test_factor_path)

fhieber replied:
good catch, thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -591,6 +652,19 @@ def _expand_nbest_translation(translation: Translation) -> List[Translation]:
+        """
+        starting_idx = min(len(target_ids), num_target_prefix_tokens)
+        target_ids = target_ids[starting_idx:]
+        return target_ids
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -183,13 +211,26 @@ def chunks(self, chunk_size: int) -> Generator['TranslatorInput', None, None]:
+                target_prefix_factors = self.target_prefix_factors
+            else:
+                target_prefix_tokens = self.target_prefix_tokens if chunk_id == 0 else None
+                target_prefix_factors = self.target_prefix_factors if chunk_id == 0 else None
```
Done. Thanks!

hoangcuong2011 commented on sockeye/test_utils.py:
```diff
@@ -58,6 +58,50 @@ def generate_digits_file(source_path: str,
 
 
+def generate_json_input_file_with_tgt_prefix(src_path:str, tgt_path: str, json_file_with_tgt_prefix_path: str, \
+                                          src_factors_path: List[str] = None, tgt_factors_path: List[str] = None, seed=13):
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -895,11 +943,20 @@ def translate(self, trans_inputs: List[TranslatorInput], fill_up_batches: bool =
                                           for translated_chunk in translations_for_input_idx]
+                if num_target_prefix_tokens > 0 and not trans_input.keep_target_prefix_key:
+                    for i in range(len(translations_to_concat)):
+                        if i == 0 or (i > 0 and trans_input.use_target_prefix_all_chunks):
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -325,8 +349,17 @@ def make_input_from_dict(sentence_id: SentenceId,
         keep_target_prefix_key = input_dict.get(C.JSON_KEEP_TARGET_PREFIX_KEY)
-        keep_target_prefix_key = False if (keep_target_prefix_key is not None and keep_target_prefix_key == "False") else True
+        if keep_target_prefix_key is None:
+            keep_target_prefix_key = True
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -325,8 +349,17 @@ def make_input_from_dict(sentence_id: SentenceId,
+
+        use_target_prefix_all_chunks = input_dict.get(C.JSON_USE_TARGET_PREFIX_ALL_CHUNKS_KEY)
+        if use_target_prefix_all_chunks is None:
+            use_target_prefix_all_chunks = True
```
Done. Thanks!

fhieber commented on sockeye/beam_search.py:
```diff
@@ -780,30 +826,53 @@ def forward(self,
-                    scores += first_step_mask
+                if target_prefix is None:
+                    scores = scores + first_step_mask if t == 1 else scores
+                else:
```
I think it would be nice to have a brief explanation here, something like:
`# while decoding target prefixes, we mask all other hypotheses than the first`

hoangcuong2011 replied:
Done. Thanks!

fhieber commented on sockeye/beam_search.py:
```diff
@@ -780,30 +826,53 @@ def forward(self,
+            if target_prefix is not None and t <= target_prefix.size(1):
+                # Make sure search selects the current prefix token by setting the scores of all
+                # other vocabulary items to infinity.
+                target_prefix_in_one_hot = utils.one_hot_encoding_from_prefix(target_prefix[:, t-1:t], self.output_vocab_size)
```
I haven't thought about this myself in depth: is there any way we can avoid the one_hot encoding, creating these large vectors?

hoangcuong2011 replied:
This is actually a very good point. I thought about it and it is easy to reflect my thoughts of how to do this by this example:
```
real target vocab, lets say tensor([ 0,  1,  2,  3,  ...., 15]) (size of (1, 16) 
vocab_slice_ids, lets say tensor([ 0,  1,  2,  3,  6,  9, 10, 12]) (size of (1, 8) 
target prefix id, lets say 9.
```
I guess the key question is to know the index where 9 places in vocab_slice_ids (5 in this example). One we get the index we can simply set it to 1 and the rest is zero. Apparently getting an index can be done by this function https://stackoverflow.com/a/47867513/1953737

I however don't know how fast it can be, especially when we need to do something like: `(t == prefix_id).nonzero(as_tuple=True)` for every prefix id. I used nonzero before and I confirm it is fast (despite nonzero and as_tuple stuff), but it is still not that clear to me that if we do that for all prefix ids in the same timestep, whether it is still that fast ...

fhieber replied:
I have another question about this: what is the current & intended behavior if vocab_slice_ids does not contain ids from the prefix?

If we give preference to the target_prefix, that is, we still force-decode the target prefix id, even though it is not part of `vocab_slice_ids`, wouldn't we need to ids from target prefix to the set of 'allowed' ids in `vocab_slice_ids`?

And then, wouldn't it make more sense to map the target_prefix_ids to the reduced vocabulary space right away before going into the beam search loop? `vocab_slice_ids` is fixed for a single request, so the 'reduced' vocabulary space is the same for each beam search iteration (which is why we cache the reduction operation in the OutputLayer).
Similarly, we could also one-hot-encode all time steps of the target prefix once before the loop, `inf`-mask it, and then only pull out the slice for each time step here. This would reduce memory allocations in each loop and be likely faster.

fhieber replied:
After thinking about this a bit more I agree that there may not be a better way to get prefix ids in the reduced space , other than going through the one-hot.

But I still think the following two changes would be useful:
- ensuring that target prefix ids are part of vocab_slice_ids
- pre-allocating a 'custom scores distribution' (similar like `eos_dist` or `pad_dist` for each target prefix timestep (for cases with and without vocabulary selection). You can do that with `x = pt.full(<shape>, value=np.inf); x[:, target_prefix_ids] = 0` Within the beam search loop, your logic would then simplify to something like `scores += x`.
- This would also allow you to put this logic into `UpdateScores`.

What do you think?

In regards to finding the index of a prefix id in the reduced space:
- ids in `vocab_slice_ids` are unique and sorted by increasing id (plus padded to a multiple of 8) 
- if we assume target prefix ids are a strict subset of `vocab_slice_ids` (we first need to make sure of that), isnt the 'reduced id' of a target prefix id simply the position in vocab_slice

hoangcuong2011 replied:
These discussions are very useful. Thanks Felix.

1. "what is the current & intended behavior if vocab_slice_ids does not contain ids from the prefix?" -> At the moment I just assume what in the prefix will be in the vocab_slice_ids. But you are right this is not great.
 
2. "ensuring that target prefix ids are part of vocab_slice_ids" -> Done I added that now.

3. "pre-allocating a 'custom scores distribution' (similar like eos_dist or pad_dist for each target prefix timestep (for cases with and without vocabulary selection). You can do that with x = pt.full(<shape>, value=np.inf); x[:, target_prefix_ids] = 0 Within the beam search loop, your logic would then simplify to something like scores += x. This would also allow you to put this logic into UpdateScores." 

I got your idea here and I need to think more about it. It is true that sum with np.inf is used with `pad_dist` and `pt.where()`. It was actually a bit difficult for me to understand these logics in the beginning but when I got it I noticed it is very smart. Meanwhile, my thought at the moment is also that I don't really know the benefit of adding sum instead of doing in-place `masked_fill_`. Doing masked_fill is intuitive, and it is an in-place operation so it is fast and I notice people do very often in Pytorch. Another thought is how to cover all cases of arbitrary length of prefix with this. Finally with the approach of using the one hot encoding method at the moment I found that we are able handle both prefix and prefix factors smoothly. I am not sure how clean it is to apply sum of np.inf for both constraining prefix and prefix factors.

Meanwhile I do notice from this discussion is that doing `one_hot_encoding_from_prefix` for each time step separately is inefficient (I did that for every time-step because I thought doing it once for the full tensor results in a tensor of size of batch_size * max_length* vocab size, which could be a big issue to memory). But since max_len of prefix is often small (I believe not a lot of people use more than, say, 5 tokens to say the least), it should be OK regarding memory usage. So I now call the function for both a full size of prefix and prefix factors. Please take a look and let me know if this is cleaner from your point of view.

hoangcuong2011 replied:
After thinking more about this, I think I agreed with you about several things:

1. `And then, wouldn't it make more sense to map the target_prefix_ids to the reduced vocabulary space right away before going into the beam search loop? vocab_slice_ids is fixed for a single request, so the 'reduced' vocabulary space is the same for each beam search iteration (which is why we cache the reduction operation in the OutputLayer).
Similarly, we could also one-hot-encode all time steps of the target prefix once before the loop, inf-mask it, and then only pull out the slice for each time step here. This would reduce memory allocations in each loop and be likely faster.` -> It is true that we should map the target_prefix_ids to the reduced vocabulary space right away before going into the beam search loop. This is a very obvious thing I should do and somehow I missed it indeed.

2. `You can do that with x = pt.full(<shape>, value=np.inf); x[:, target_prefix_ids] = 0 Within the beam search loop, your logic would then simplify to something like scores += x.` I agreed that while mask_filled_ works perfectly fine, it is better here to change it as simple as `scores += x`. It is better because of at least two reasons (1): it looks simpler indeed. and (2): it is consistent because it follows our practice of similar stuff (+ first_step_mask, + scores in _update_scores). Furthermore, the logic here can be applied exactly to prefix factors masking. So yeah I like this idea much more now. So I changed our code based on your suggestion.

BTW Regarding (2): I also changed the generate_one_hot_function to `gen_prefix_masking` instead, as we now create a new matrix which are inf everywhere except zero for prefix ids. But the logic is still the same for that function (i.e. gen_prefix_masking) because it is based on the pt.nn.functional.one_hot. So everything is ``smooth" here.

Meanwhile I still prefer keeping +x separately instead of putting it into UpdateScores at the moment. There are two reasons here. 1. In greedy module we don't have `updatescores` stuff. So keeping the same line of code (+x) between two modules makes things consistency. 2. tracing `UpdateScores` seems does not gain if not hurt on GPUs so at least there is no need yet to add stuff into UpdateScores at the moment for a better performance.

fhieber commented on sockeye/inference.py:
```diff
@@ -591,6 +644,18 @@ def _expand_nbest_translation(translation: Translation) -> List[Translation]:
     return nbest_list
 
+def _remove_target_prefix_tokens(target_ids: TokenIds,
+                     num_target_prefix_tokens: int) -> TokenIds:
```
[nit] strange indent

hoangcuong2011 replied:
Done. Thanks!

fhieber commented on test/common.py:
```diff
@@ -135,6 +175,39 @@ def test_scoring(data: Dict[str, Any], translate_params: str, test_similar_score
+        for json_output in data['test_with_target_prefix_outputs']:
+            print(json_output['translation'], file=target_out)
+            for i, factor_out in enumerate(target_factor_outs, 1):
+                factor = json_output['factor%d' % i]
```
format string

hoangcuong2011 replied:
Done. Thanks!

fhieber reviewed (commented):
Thanks for iterating!

fhieber commented:
> Thank you for your great feedbacks. I revised the code to address all of them. Please review it again.
> 
> About Felix's point: ``I wonder though if there would be value in thinking about how we could implement this with code from the scoring modules, the advantage being that computation over time steps within the target prefix could be parallelized." This is a great thing to think of indeed. I thought about it. The good thing is as you mention we can parallel the computation. However, there are also important disadvantages:
> 
> * Implementing by constraining the beam like what we are doing matches with the way we are implementing decoding. Thus the code is easy to read and understand. If we implement in the parallelism way, it breaks the matching in the logic of the implementation to some degree I think.
> * The length of prefix in MT is not as long as the prefix that is used, for instance, in GPT3. I think for MT it is about 0-1-2 tokens at most. So the benefit of parallelism could be marginal (but let's see how our ``customers" use the features, maybe they want to use long prefix)
> * By implementing like we are doing, our code is very flexible to handle sentences in the same batch with different target prefixes (e.g. one has zero prefix, one has one prefix token, one has 2 prefix token). Implementing stuff in parallel makes it very nontrivial to handle that flexibility.

Thanks for elaborating! Especially your last point convinced me; lets go with the current implementation.

fhieber commented on sockeye/inference.py:
```diff
@@ -183,13 +211,22 @@ def chunks(self, chunk_size: int) -> Generator['TranslatorInput', None, None]:
+            # otherwise target_prefix_tokens are assigned only to the first chunk
+            target_prefix_tokens = self.target_prefix_tokens if chunk_id == 0 or self.use_target_prefix_all_chunks else None
+            target_prefix_factors = self.target_prefix_factors if chunk_id == 0 or self.use_target_prefix_all_chunks else None
+            keep_target_prefix_key = self.keep_target_prefix_key
```
You don't need this line

hoangcuong2011 replied:
Done. Thanks!

fhieber commented on sockeye/inference.py:
```diff
@@ -205,6 +242,9 @@ def with_eos(self) -> 'TranslatorInput':
                                source_prefix_factors=self.source_prefix_factors,
+                               target_prefix_tokens=self.target_prefix_tokens,
+                               target_prefix_factors=self.target_prefix_factors,
+                               use_target_prefix_all_chunks=self.use_target_prefix_all_chunks,
```
This is missing the assignment/setting for `keep_target_prefix_key`, right?

fhieber commented on sockeye/test_utils.py:
```diff
@@ -58,6 +58,50 @@ def generate_digits_file(source_path: str,
+                src_digits, tgt_digits = stdigits[0].strip(), stdigits[1].strip()
+                tgt_prefix = tgt_digits.split()
+                if len(tgt_digits) > 0:
+                    random_pos = random_gen.choice([pos for pos in range(len(tgt_prefix))])
```
This can give you empty prefixes, right? What is the intended behavior then? Should we even allow this?

hoangcuong2011 replied:
This can give you empty prefixes, right? -> Yes
What is the intended behavior then? -> The intended behavior is to treat it like there is no prefix.
Should we even allow this? -> I don't know indeed Felix. I just thought: 

1 - it is very easy to handle that case, no need to adhoc modify anything.
2- No need for logger.error (Oops you wanted to use a prefix but the prefix is actually an empty string). 

Let me know if you have a different opinion.

fhieber replied:
If empty prefixes work with the current implementation, thats great. But maybe we can add a logger.warning somewhere in the `TranslatorInput` construction code. A non-None prefix tensor will trigger your implementation in this PR to some extent. I don't think it would have a large effect on performance, but we should still warn the user, it may not be intended. What do you think?

hoangcuong2011 replied:
Done. You are right we should add warning for empty string of prefix token and empty list of prefix factors.

tdomhan commented on docs/inference.md:
````diff
@@ -96,6 +96,36 @@ Similar to source factors, source prefix factors can be also specified with JSON
+If your model was trained with target factors, you can add target prefix factors during inference with JSON format as follows:
+
+```json
+{ "text": "The boy ate the waff@@ le .", "target_prefix_factors": ["O"]}
````
+1 if this is supported you could add an explanation for how this works. Namely, if there's only a factor what will the corresponding token be?

tdomhan replied:
Also maybe it should be clarified what the semantics of target prefix are, namely this is a force-decoded first target token (and not a replacement for the `<bos>` token), right?

hoangcuong2011 replied:
Yes it also supports to add target factors only.

hoangcuong2011 replied:
Thanks a lot Tobias. Following your suggestions I updated inference.md to make it more clear.

hoangcuong2011 commented on CHANGELOG.md:
```diff
@@ -11,6 +11,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
+
+### Added
+- Added support for the use of adding target prefix to the input in JSON format during inference.
+- Added support for the use of adding target prefix factors to the input in JSON format during inference.
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -79,10 +81,15 @@ def decode_step(self,
                     tf_logits = pt.log_softmax(tf_logits, dim=-1)
                 tf_scores = -tf_logits
+                if target_prefix_factors is not None:
+                    beam_size = tf_scores.size()[0]//target_prefix_factors.size()[0]
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -79,10 +81,15 @@ def decode_step(self,
                     tf_logits = pt.log_softmax(tf_logits, dim=-1)
                 tf_scores = -tf_logits
+                if target_prefix_factors is not None:
+                    beam_size = tf_scores.size()[0]//target_prefix_factors.size()[0]
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -79,10 +81,15 @@ def decode_step(self,
         if target_factor_outputs:
             predictions = []  # type: List[pt.Tensor]
-            for tf_logits in target_factor_outputs:
+            for i, tf_logits in enumerate(target_factor_outputs):
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -138,6 +146,12 @@ def decode_step(self,
                 target_factor_probs = [tfo.softmax(dim=-1) for tfo in target_factor_outputs]
+                if target_prefix_factors is not None:
+                    for i in range(len(target_factor_probs)):
+                        beam_size = int(target_factor_probs[i].size()[0]/target_prefix_factors.size()[0])
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -532,12 +546,14 @@ def __init__(self,
                  num_target_factors: int,
-                 inference: _SingleModelInference):
+                 inference: _SingleModelInference,
+                 output_vocab_size: int):
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -559,6 +577,7 @@ def forward(self,
         :param restrict_lexicon: Lexicon to use for vocabulary restriction.
         :param max_output_lengths: ndarray of maximum output lengths per input in source.
                 Shape: (batch_size=1,). Dtype: int32.
+        :param target_prefix: Target prefix ids. Shape: (batch_size=1, max target prefix length).
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -585,10 +604,31 @@ def forward(self,
 
         t = 1
+        if target_prefix_factors is not None and C.TARGET_FACTOR_SHIFT:
+            target_prefix_factors_shift = pt.zeros(target_prefix_factors.size(0), target_prefix_factors.size(1) + 1, target_prefix_factors.size(2)).to(self.device).type(target_prefix_factors.dtype)
```
Done (both avoid repeatedly asking for size() and specify dtype in pt.zeros). Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -585,10 +604,31 @@ def forward(self,
         # TODO: check for disabled predicted output length
 
         t = 1
+        if target_prefix_factors is not None and C.TARGET_FACTOR_SHIFT:
```
Thanks. 
I created a utils.shift_prefix_factors function for this and call it from _get_inference_input() to avoid calling it in different places (between beamsearch and greedysearch).

`I target_prefix_factors = utils.shift_prefix_factors(target_prefix_factors) if target_prefix_factors is not None and C.TARGET_FACTOR_SHIFT else target_prefix_factors
`

Here is also the comment to explain while we need to shift prefix factors:

`# During inference, if C.TARGET_FACTOR_SHIFT is True, predicting target_factors are left-shifted (see _unshift_target_factors function()) so that they re-align with the words. With that, target_prefix_factors need to be also right-shifted here if C.TARGET_FACTOR_SHIFT is True so that when they are shifted back later they would align with words.
`

hoangcuong2011 commented on sockeye/utils.py:
```diff
@@ -264,6 +264,21 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+    :return prefix_in_one_hot (batch size, 1, output_vocab_size)
+
+    """
+    prefix_in_one_hot = pt.nn.functional.one_hot(prefix.to(pt.int64), num_classes=vocab_size)
```
This is interesting. I am not that sure (not clear in the document from pytorch)... To be safe let us also assign the device of prefix_in_one_hot to prefix.device as well to avoid any potential error.

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -780,30 +826,53 @@ def forward(self,
         estimated_reference_lengths = estimated_reference_lengths.repeat_interleave(self.beam_size, dim=0)
 
         t = 1
+        if target_prefix_factors is not None and C.TARGET_FACTOR_SHIFT:
```
Done (see above comments). Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -864,6 +933,93 @@ def forward(self,
+                         [1 1 0]                       [inf inf 0]
+        """
+        # Step 1
+        masking = pt.zeros((target_prefix.size(0), target_prefix.size(1) + 1)).to(self.device)
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -906,7 +1062,8 @@ def get_search_algorithm(models: List[SockeyeModel],
                                             skip_softmax=True,
-                                            constant_length_ratio=0.0))
+                                            constant_length_ratio=0.0),
+            output_vocab_size=models[0].output_layer_vocab_size)
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -132,14 +132,18 @@ class TranslatorInput:
     def __str__(self):
-        return 'TranslatorInput(%s, %s, factors=%s, source_prefix_tokens=%s, source_prefix_factors=%s, constraints=%s, avoid=%s)' \
-            % (self.sentence_id, self.tokens, self.factors, self.source_prefix_tokens, self.source_prefix_factors, self.constraints, self.avoid_list)
+        return 'TranslatorInput(%s, %s, factors=%s, source_prefix_tokens=%s, source_prefix_factors=%s, target_prefix_tokens=%s, target_prefix_factors=%s, use_target_prefix_all_chunks=%s, keep_target_prefix_key=%s, constraints=%s, avoid=%s)' \
```
Done (reformat and f'). Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -864,6 +933,93 @@ def forward(self,
                             lengths=lengths,
                             estimated_reference_lengths=estimated_reference_lengths)
 
+    def adjust_first_step_masking(self,
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -888,24 +962,30 @@ def translate(self, trans_inputs: List[TranslatorInput], fill_up_batches: bool =
         # assembling source ids on cpu array (faster) and copy to Translator.device (potentially GPU) in one go below.
         source = onp.zeros((batch_size, max_length, self.num_source_factors), dtype='int32')
 
+        target_prefix = onp.zeros((batch_size, max_target_prefix_length), dtype='int32') if max_target_prefix_length > 0 else None
```
Yeah that is the reason. it is also that target prefix and target prefix factors could have different lengths. Keeping target_prefix and  target_prefix_factors separate makes this to be easily/effectively supported.

I don't know how often our ``customers" want to use target_prefix_factors only, or use target_prefix_factors with different length to the target_prefix though. I just hope the flexibility here will help them.

fhieber replied:
Right, there could be some use to it, depending on what users define as target factors. Let's have this flexibility.

hoangcuong2011 commented on test/common.py:
```diff
@@ -95,14 +111,37 @@ def test_translate_equivalence(data: Dict[str, Any], translate_params_equiv: str
+                prefix = json_output_with_target_prefix['target_prefix_factors']
+                if len(prefix) > 0:
+                    for j in range(1, len(prefix) + 1):
+                        factors_from_translation = json_output_with_target_prefix['factor' + str(j)]
```
Done. Thanks!

hoangcuong2011 commented on test/common.py:
```diff
@@ -135,6 +175,39 @@ def test_scoring(data: Dict[str, Any], translate_params: str, test_similar_score
                 print(factor, file=factor_out)
 
+    target_with_target_prefix_path = os.path.join(data['work_dir'], "score_with_target_prefix.target")
+    target_with_target_prefix_factor_paths = [os.path.join(data['work_dir'], "score_with_target_prefix.target.factor%d" % i) for i, _ in
```
Done. Thanks!

hoangcuong2011 commented on sockeye/beam_search.py:
```diff
@@ -780,30 +826,53 @@ def forward(self,
+                                                                                         model_states,
+                                                                                         vocab_slice_ids)
+            else:
+                target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
```
Is it what you meant? 

Change the code from

```
if target_prefix_factors is None:
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids)
else:
    target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids,
                                                                             target_prefix_factor)
```
to

```
target_prefix_factor = target_prefix_factors[:, t-1, :] if \
                       target_prefix_factors is not None \
                       and self.num_target_factors > 1 \
                       and t <= target_prefix_factors.size(1) else None

target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                         model_states,
                                                                         vocab_slice_ids,
                                                                         target_prefix_factor)
```

fhieber replied:
yes exactly

hoangcuong2011 replied:
Somehow if I slightly changed the code from:

```
if target_prefix_factors is None:
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids)
else:
    target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids,
                                                                             target_prefix_factor)
```

to 
```
if target_prefix_factors is None:
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids, None)
else:
    target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids,
                                                                             target_prefix_factor)
```

It does not pass the test_beam_search.py. The specific error is as follows:
```

test/unit/test_beam_search.py:326: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/torch/nn/modules/module.py:1102: in _call_impl
    return forward_call(*input, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = BeamSearch(
  (_repeat_states): RepeatStates()
  (_sort_states): SortStates()
  (_update_scores): UpdateScores()
  (_s...(_lp): LengthPenalty()
    )
  )
  (_top): TopK()
  (_traced_repeat_states): RepeatStates(original_name=RepeatStates)
)
source = tensor([[[2.],
         [4.],
         [3.],
         [0.],
         [0.]]]), source_length = tensor([3])
restrict_lexicon = None, max_output_lengths = tensor([3], dtype=torch.int32), target_prefix = None, target_prefix_factors = None




        t = 1
        for t in range(1, max_iterations + 1):  # max_iterations + 1 required to get correct results
            # (1) obtain next predictions and advance models' state
            # target_dists: (batch_size * beam_size, target_vocab_size)
            # target_factors: (batch_size * beam_size, num_secondary_factors, 2),
            # where last dimension holds indices and scores
            if target_prefix_factors is None:
>               target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                                         model_states,
                                                                                         vocab_slice_ids, None)
E               TypeError: decode_step() takes from 3 to 4 positional arguments but 5 were given

sockeye/beam_search.py:841: TypeError
```

Do you know why?

fhieber replied:
test_beam_search uses a `_TestInference` helper class: https://github.com/awslabs/sockeye/blob/main/test/unit/test_beam_search.py#L258 
You need to add the additional argument there as well.

hoangcuong2011 replied:
Thanks. I added "`, *args`" as additional argument to that function. Then I also changed the way we add  target_prefix_factor into decode_step from

```
if target_prefix_factors is None:
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids)
else:
    target_prefix_factor = target_prefix_factors[:, t-1, :] if self.num_target_factors > 1 and t <= target_prefix_factors.size(1) else None
    target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                             model_states,
                                                                             vocab_slice_ids,
                                                                             target_prefix_factor)
```
to

```
target_prefix_factor = target_prefix_factors[:, t-1, :] if target_prefix_factors is not None \
                                                       and self.num_target_factors > 1 \
                                                       and t <= target_prefix_factors.size(1) \
                                                       else None
target_dists, model_states, target_factors = self._inference.decode_step(best_word_indices,
                                                                         model_states,
                                                                         vocab_slice_ids,
                                                                         target_prefix_factor)
```
It looks cleaner indeed! :-)

fhieber commented on sockeye/beam_search.py:
```diff
@@ -67,10 +68,17 @@ def encode_and_initialize(self, inputs: pt.Tensor, valid_length: Optional[pt.Ten
         states, predicted_output_length = self._model.encode_and_initialize(inputs, valid_length, self._const_lr)
         return states, predicted_output_length
 
+    def get_model(self):
```
It seems you are working around the fact that we made the model a 'private' member of this class by naming it `_model`. You can change that if you want, then you don't need this getter (which isn't really pythonic either).
Alternatively you can do:
```
@property
def model(self):
  return self._model
```
and access via `inference.model`

hoangcuong2011 replied:
Done. Thanks!

hoangcuong2011 commented:
Thank you Felix for another careful/thoughtful iteration of reviewing and Tobias for adding good suggestions of revising inference.md better. I addressed all of your comments. Please let me know if you have additional comments.

fhieber commented on sockeye/utils.py:
```diff
@@ -264,6 +264,124 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+    """
+    prefix_in_one_hot = pt.nn.functional.one_hot(prefix.to(pt.int64), num_classes=vocab_size).to(prefix.device)
+    # all hots are assigned to 1 if a prefix id is 0 (i.e. no prefix)
+    prefix_in_one_hot.masked_fill_(prefix.unsqueeze(-1) == 0, 1)
```
I don't think I understand this line. Isn't the one-hot from line 276 already only 0s and 1s? I assume `prefix` is `(batch size, 1)` and looks like this: 
```
prefix = pt.tensor([1,2,3]).unsqueeze(1)
```
What does this line do then?

hoangcuong2011 replied:
Thanks for this. I definitely should add more explanation regarding this. So now I added a detailed explanation there. Pls let me know if it is still not that clear to you.

```
    # In the same batch during inference, it is possible that some translations have target prefix
    # while others do not have. It is also possible that translation may have a target prefix with
    # different length to others. Thus prefix ids may include a full zero vector if a translation
    # in the batch does not have prefix, or include a vector padding with zeros on the right if some
    # translations are with shorter prefix. An example of prefix ids reflecting length differences \
    # is as follows:
    #
    # [1, 2, 3]
    # [1, 2, 0]
    # [0, 0, 0]
    #
    # Here, the first sentence has a prefix of length 3, the second one has a prefix of length 1 \
    # and the last one does not have prefix.
    #
    # At any timestep, some target prefix ids could be zero. If a prefix id is zero for a translation \
    # at a timestep, all hots in the vocab are assigned to 1 (instead of only one hot is assigned to 1). \
    # This makes sure there is no constraint on selecting any specific target token for the translation
    # in that case.

    prefix_in_one_hot.masked_fill_(prefix.unsqueeze(-1) == 0, 1)
```

fhieber replied:
oh I see, the semantics of `0` / `C.PAD_ID` in the `target_prefix` weren't clear to me. `0` in the `target_prefix` means 'no constraint'

fhieber reviewed (commented):
Thanks @hoangcuong2011 ! I have a few more comments, thanks for iterating -- we are getting there :)

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -591,6 +651,16 @@ def _expand_nbest_translation(translation: Translation) -> List[Translation]:
 
     return nbest_list
 
```
Done. Thanks!

hoangcuong2011 commented on sockeye/inference.py:
```diff
@@ -951,7 +1041,15 @@ def _get_inference_input(self,
+        target_prefix = pt.tensor(target_prefix, device=self.device, dtype=pt.int32) if target_prefix is not None else None  # type: ignore
+        target_prefix_factors_tensor = pt.tensor(target_prefix_factors, device=self.device, dtype=pt.int32) if target_prefix_factors is not None else None  # type: ignore
+
+        # During inference, if C.TARGET_FACTOR_SHIFT is True, predicting target_factors are left-shifted (see _unshift_target_factors function()) so that \
```
Done. Thanks!

hoangcuong2011 commented on sockeye/utils.py:
```diff
@@ -264,6 +264,124 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+    :return prefix_in_one_hot (batch size, 1, output_vocab_size)
+
+    """
+    prefix_in_one_hot = pt.nn.functional.one_hot(prefix.to(pt.int64), num_classes=vocab_size).to(prefix.device)
```
I honestly don't know.... There is no document on the internet on one_hot function regarding of the device of its resulting tensor. So I don't know whether prefix_in_one_hot is on the same device to prefix. That is why I assign to prefix.device for safety ....

fhieber replied:
ok, thanks!

hoangcuong2011 commented on sockeye/utils.py:
```diff
@@ -264,6 +264,124 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+                         [1 1 0]                       [inf inf 0]
+                         [1 1 0]                       [inf inf 0]
+        """
+        first_step_mask_sizes = first_step_mask.size()
```
Done, thanks!

hoangcuong2011 commented on sockeye/utils.py:
```diff
@@ -264,6 +264,124 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+        # Step 1
+        masking = pt.zeros((target_prefix_sizes[0], target_prefix_sizes[1] + 1), device=target_prefix.device)
+        masking[:,:target_prefix_sizes[1]] = target_prefix
+        masking.masked_fill_(masking != 0., 1.0)
```
I am not quite sure whether to go with clamp function because it is not that intuitive to me how to use ... I googled and found not that easy to figure it out a sequence of min and max formula. Not sure what others think.

pt.where is a good option because it is much easier to understand compared to clamp. However, one nicer thing with masked_fill_ is that it is an in-place operation....

fhieber replied:
clamp just limits the minimum and maximum value, which is what you want here, right? To me the benefit of `clamp` over masking would be that with `clamp` you don't need to explicitly create the full mask of shape `masking.size()`. 

I timed this locally in my laptop and clamp seems to be a bit faster.

If we'd ever want to script/trace that function would inplace still be a benefit?

hoangcuong2011 replied:
You are right. What I wanted here is simply forcing all non zero ids to 1, which is also equivalent to limiting the minimum and maximum value 0 and 1. I got it now.

I actually also tried to masked_fill_ and and clamp and it is true that clamp is faster on both CPU and GPU (I am very surprised to see this, an interesting learning for me :D). So yeah lets go for clamp indeed. I just changed it. Thanks for the suggestion!


```
>>> import timeit
>>> 
>>> set_up_line = """
... import torch as pt
... pt.manual_seed(1)
... masking = pt.zeros((10, 10))"""
>>> 
>>> executive_line = """
... masking[:,:7] = pt.randint(0, 10, (10, 7))
... masking.masked_fill_(masking != 0., 1.)
... """
>>> 
>>> timeit.timeit(executive_line,setup=set_up_line, number=100000)

2.1035887990146875
>>> 
>>> executive_line = """
... masking[:,:7] = pt.randint(0, 10, (10, 7))
... masking = pt.clamp(masking, 0., 1.)
... """
>>> timeit.timeit(executive_line,setup=set_up_line, number=100000)
1.3465280530508608
```

fhieber commented on sockeye/inference.py:
```diff
@@ -303,6 +350,19 @@ def make_input_from_dict(sentence_id: SentenceId,
                 return _bad_input(sentence_id, reason=str(input_dict))
 
+        target_prefix_tokens = input_dict.get(C.JSON_TARGET_PREFIX_KEY)
+        if target_prefix_tokens is not None and not target_prefix_tokens:
```
if the target_prefix is a single space (for whatever reason) (`' '`) this will not fire but the target_prefix_tokens will still be empty. How about moving the check into line 357, after the `get_tokens` call:
```
if not target_prefix_tokens:   # none or empty list
   logger.warning(f"Empty string is specified as a target prefix for input '{input_dict[C.JSON_TEXT_KEY]}'.")
```

hoangcuong2011 replied:
Thanks. I moved the check into line 357 as you suggested and slightly changed it to:

```
target_prefix_tokens = input_dict.get(C.JSON_TARGET_PREFIX_KEY)
target_prefix_tokens = list(utils.get_tokens(target_prefix_tokens)) if target_prefix_tokens is not None else None
if target_prefix_tokens is not None and not target_prefix_tokens:
    logger.warning(f"Empty string is specified as a target prefix for input '{input_dict[C.JSON_TEXT_KEY]}'.")
```
This avoid giving us `warning` when `target_prefix_tokens` is None, which is the case when a ``user" does not specify `JSON_TARGET_PREFIX_KEY` in the JSON input.

fhieber replied:
thanks!

hoangcuong2011 commented:
Thank you so much @fhieber for another round of careful reviewing and with great suggestions. 
I addressed all of them with the latest iteration, please take a look. 

Let me know if anything is not clear to think or anything you think it could be improved.

tdomhan commented on sockeye/beam_search.py:
```diff
@@ -495,9 +523,13 @@ def forward(self, best_hyp_indices, *states):
     vocab_slice_ids = restrict_lexicon.get_trg_ids(source_words.cpu().int().numpy())
+    if target_prefix is not None:
+        # Ensuring that target prefix ids are part of vocab_slice_ids
+        vocab_slice_ids = onp.lib.arraysetops.unique(onp.concatenate([vocab_slice_ids, target_prefix.cpu().int().numpy().flatten()]))
```
Alternatively, you could first convert `vocab_slice_ids` to pytorch and use pytorch's unique.

hoangcuong2011 replied:
Thanks Tobias for the suggestion. I tried and noticed a nice improvement in speed (from 6.0s to 5.4s in my test, around 10% improvement) by converting vocab_slice_ids to pytorch and use pytorch's unique function. Yeah, so let's go for it.

Testing code for speed:
```
import timeit
set_up_line = """
import numpy as np
import torch as pt
pt.manual_seed(1)
np.random.seed(1)
vocab_slice_idss = np.random.randint(0, 10, (100, 50))
target_prefix = pt.randint(0, 10, (100, 50))
"""
executive_line = """
z = target_prefix.device
for i in range(target_prefix.size(0)):
    vocab_slice_ids = np.lib.arraysetops.unique(np.concatenate([vocab_slice_idss[i], target_prefix[i].cpu().int().numpy().flatten()]))
    vocab_slice_ids = pt.nn.functional.pad(pt.tensor(vocab_slice_ids, dtype=pt.int64, device=z),  # type: ignore
                                           pad=(0, 7 - ((vocab_slice_ids.size - 1) % 8)),
                                           mode='constant', value=2)
"""
timeit.timeit(executive_line,setup=set_up_line, number=1000)
6.065197644056752


executive_line = """
z = target_prefix.device
for i in range(target_prefix.size(0)):
	vocab_slice_ids = pt.tensor(vocab_slice_idss[i])
	vocab_slice_ids = pt.unique(pt.concat([vocab_slice_ids.type(pt.int64).to(z), target_prefix[0].type(pt.int64).to(z)], -1))
	vocab_slice_ids = pt.nn.functional.pad(vocab_slice_ids.clone().detach(),  # type: ignore
	                                           pad=(0, 7 - ((vocab_slice_ids.size(-1) - 1) % 8)),
	                                           mode='constant', value=2)
                                           """
timeit.timeit(executive_line,setup=set_up_line, number=1000)
5.411461449926719
```

tdomhan replied:
Great, thanks for benchmarking!

tdomhan commented on sockeye/beam_search.py:
```diff
@@ -70,19 +71,32 @@ def encode_and_initialize(self, inputs: pt.Tensor, valid_length: Optional[pt.Ten
+            target_prefix_factor_masks = None  # type: Optional[pt.Tensor]
+            factor_vocab_size = target_factor_outputs[0].size(-1)  # type: ignore
+            if target_prefix_factor is not None:
+                batch_beam, _ = scores.size()
```
I wonder if it would make sense to expand `target_prefix_factor` to batch*beam already in GreedySearch/BeamSearch where you know the batch/beam size and don't need to calculate it. What do you think?

hoangcuong2011 commented:
Hi, just a heads up: I went another iteration following suggestions from Tobias:
1. Completely expand target_prefix_factor to to batch*beam already in GreedySearch/BeamSearch. This helps avoid computing batch, beam inside within decoding_step. 
2. Convert vocab_slice_ids to pytorch and use pytorch's unique function (i.e using pytorch functions to make sure prefix is always include in restricted vocab). By doing that I noticed it is faster around 10%.
Thanks Tobias!

fhieber commented on sockeye/beam_search.py:
```diff
@@ -46,7 +46,17 @@ def encode_and_initialize(self,
                     states: List,
-                    vocab_slice_ids: Optional[pt.Tensor] = None):
+                    vocab_slice_ids: Optional[pt.Tensor] = None,
+                    target_prefix_factor: Optional[pt.Tensor] = None,
```
Needs to be renamed to `target_prefix_factor_mask`

hoangcuong2011 replied:
Done. Thanks!

fhieber commented on sockeye/beam_search.py:
```diff
@@ -495,13 +529,23 @@ def forward(self, best_hyp_indices, *states):
+        # Ensuring that target prefix ids are part of vocab_slice_ids
+        vocab_slice_ids = pt.unique(pt.concat([pt.tensor(vocab_slice_ids_np, device=device, dtype=pt.int64), pt.flatten(target_prefix).type(pt.int64)], -1))
+        # Pad to a multiple of 8.
+        vocab_slice_ids = pt.nn.functional.pad(vocab_slice_ids.clone().detach(),
```
just curious: why do you need the `clone()` and `detach()`?

fhieber replied:
But why do you need to copy in the first place? You are not calling `pt.tensor` here. 
In any case, I think you could simplify this whole block to:
```
vocab_slice_ids = pt.tensor(vocab_slice_ids_np, device=device, dtype=pt.int64)
if target_prefix is not None:
    # Ensuring that target prefix ids are part of vocab_slice_ids
     vocab_slice_ids = pt.concat([vocab_slice_ids, target_prefix.flatten().type(pt.int64)], -1).unique()
# Pad to a multiple of 8.
vocab_slice_ids = pt.nn.functional.pad(vocab_slice_ids, pad=(0, 7 - ((vocab_slice_ids_np.size - 1) % 8)), mode='constant', value=eos_id)
```

hoangcuong2011 replied:
Haha silly me. Done. Very true and thanks Felix!

fhieber replied:
thanks! I still think you can also avoid the duplicate call to `pad` as suggested above. After that I think this PR looks good to me! :)

hoangcuong2011 replied:
Yes, your code is very succinct. Done, thanks!

fhieber commented on sockeye/inference.py:
```diff
@@ -869,11 +948,20 @@ def translate(self, trans_inputs: List[TranslatorInput], fill_up_batches: bool =
         chunks_by_input_idx = itertools.groupby(translated_chunks, key=lambda translation: translation.input_idx)
         for trans_input, (input_idx, translations_for_input_idx) in zip(trans_inputs, chunks_by_input_idx):
             translations_for_input_idx = list(translations_for_input_idx)  # type: ignore
+            num_target_prefix_tokens = trans_input.num_target_prefix_tokens()
```
[nit] If you make `TranslatorInput.num_target_prefix_tokens` a `@property` you can save this line and just use `trans_input.num_target_prefix_tokens` directly below.

fhieber replied:
makes sense, thank you!

fhieber reviewed (commented):
Just a few minor remaining comments. Looks really good now!

fhieber reviewed (approved):
Again, thanks for iterating. Code looks really good now! This is a very useful feature that adds a lot of functionality to Sockeye with limited additional complexity.

tdomhan commented on sockeye/inference.py:
```diff
@@ -132,17 +132,28 @@ class TranslatorInput:
-        return 'TranslatorInput(%s, %s, factors=%s, source_prefix_tokens=%s, source_prefix_factors=%s, constraints=%s, avoid=%s)' \
-            % (self.sentence_id, self.tokens, self.factors, self.source_prefix_tokens, self.source_prefix_factors, self.constraints, self.avoid_list)
+        return f'TranslatorInput({self.sentence_id}, {self.tokens}, factors={self.factors}, \
+                                  source_prefix_tokens={self.source_prefix_tokens}, \
```
How does this actually look like? I'm asking as it seems like the newline would be included in the string, i.e. are the outputs like this (i.e. lines with a lot of leading whitespace)?
```
TranslatorInput({self.sentence_id}, {self.tokens}, factors={self.factors},
                                  source_prefix_tokens={self.source_prefix_tokens},
```
?

For a long multi-line string one could do:
```
        return (f'TranslatorInput({self.sentence_id}, {self.tokens}, factors={self.factors}, 
                                  f'source_prefix_tokens={self.source_prefix_tokens}', 
                                  f'source_prefix_factors={self.source_prefix_factors}', 
                                  f'target_prefix_tokens={self.target_prefix_tokens}', 
                                  f'target_prefix_factors={self.target_prefix_factors}', 
                                  f'use_target_prefix_all_chunks={self.use_target_prefix_all_chunks}', 
                                  f'keep_target_prefix_key={self.keep_target_prefix_key}', 
                                  f'constraints={self.constraints}', 
                                  f'avoid={self.avoid_list})')
```

Or if the intention is a multi-line string I'd reduce the whitespace indentation.

hoangcuong2011 replied:
Thanks,
You are right it is with lots of leading spaces. I tested and got something like this: 
`TranslatorInput(36, ['9', '0', '0', '2', '6', '6', '6', '1'], factors=[['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l'], ['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l'], ['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l']],                                   source_prefix_tokens=None,                                   source_prefix_factors=None,                                   target_prefix_tokens=['0', '0', '1', '2', '6', '6'],                                   target_prefix_factors=[[]],                                   use_target_prefix_all_chunks=True,                                   keep_target_prefix_key=True,                                   constraints=None,                                   avoid=None)
`
I changed the code as you suggested. Now it gives good output (no unnecessary extra space). Thanks!
`TranslatorInput(36, ['9', '0', '0', '2', '6', '6', '6', '1'], factors=[['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l'], ['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l'], ['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l']]", 'source_prefix_tokens=None', 'source_prefix_factors=None', "target_prefix_tokens=['0', '0', '1', '2', '6', '6']", 'target_prefix_factors=[[]]', 'use_target_prefix_all_chunks=True', 'keep_target_prefix_key=True', 'constraints=None', 'avoid=None)')
`

hoangcuong2011 replied:
OK Pylint does not like having multiple lines of texts.

To make it simple let us put everything in one single line instead. This is not perfect (very long line) but no need of extra f and Pylint likes it more (than splitting multiple lines.)

`return f'TranslatorInput({self.sentence_id}, {self.tokens}, factors={self.factors}, source_prefix_tokens={self.source_prefix_tokens}, source_prefix_factors={self.source_prefix_factors}, target_prefix_tokens={self.target_prefix_tokens}, target_prefix_factors={self.target_prefix_factors}, use_target_prefix_all_chunks={self.use_target_prefix_all_chunks}, keep_target_prefix_key={self.keep_target_prefix_key}, constraints={self.constraints}, avoid={self.avoid_list})'`

The output is also good, I confirmed. Something like:
`TranslatorInput(36, ['9', '0', '0', '2', '6', '6', '6', '1'], factors=[['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l'], ['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l'], ['h', 'l', 'l', 'l', 'h', 'h', 'h', 'l']]", 'source_prefix_tokens=None', 'source_prefix_factors=None', "target_prefix_tokens=['0', '0', '1', '2', '6', '6']", 'target_prefix_factors=[[]]', 'use_target_prefix_all_chunks=True', 'keep_target_prefix_key=True', 'constraints=None', 'avoid=None)')`

tdomhan replied:
oh ok. Thanks for checking!

tdomhan commented on test/common.py:
```diff
@@ -155,9 +228,11 @@ def test_scoring(data: Dict[str, Any], translate_params: str, test_similar_score
     if test_similar_scores:
-        for inp, translate_json, score_scores in zip(data['test_inputs'],
+        for inp, translate_json, translate_with_target_prefix_json, score_scores, score_with_target_prefix_scores in zip(data['test_inputs'],
                                                      data['test_outputs'],
```
nit: no longer aligned to `zip(`, maybe just move the first argument of zip to the next line.

hoangcuong2011 replied:
Done. Thanks!

tdomhan reviewed (approved):
Thanks for iterating! Leaving more further minor comments, but I don't want to stand in the way of merging.

hoangcuong2011 commented on sockeye/utils.py:
```diff
@@ -264,6 +264,148 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+        masking[:, 0] = 1.
+
+        # Step 2
+        masking = masking.unsqueeze(1).expand(-1, beam_size, -1).reshape(batch_beam, -1)
```
Done, thanks!

hoangcuong2011 commented on sockeye/utils.py:
```diff
@@ -264,6 +264,148 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+    # and other hots are inf). This makes sure there is no constraint on selecting any specific target \
+    # token for the translation in that case.
+
+    prefix_masks.masked_fill_(prefix.unsqueeze(-1) == 0, 0)
```
Thanks Tobias. I think put 0 instead of PAD makes it more clear as 0 in the target_prefix explicitly means 'no constraint'. 

Take your suggestion into account I added more detail in the comment (before this line of code):

```
# At any timestep, some target prefix ids could be 0 (i.e. 0 in the target_prefix means 'no constraint'). \
# If a prefix id is 0 for a translation at a timestep, all hots in the vocab are assigned to 0 (instead \
# of only one hot is assigned to 0 and other hots are inf). This makes sure there is no constraint on \
# selecting any specific target token for the translation in that case.

prefix_masks.masked_fill_(prefix.unsqueeze(-1) == 0, 0)
```

fhieber replied:
thanks!

tdomhan commented on sockeye/utils.py:
```diff
@@ -264,6 +264,148 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+
+    # prefix_masks are inf everywhere except zero for indices of prefix ids.
+    prefix_masks = pt.full(prefix_one_hot.size(), fill_value=np.inf, device=prefix.device, dtype=dtype)
+    prefix_masks[prefix_one_hot != 0] = 0.
```
great. Glad it improves speed.
